runtime/parachains/src/ump.rs

```diff
@@ -461,51 +460,63 @@ impl<T: Config> Pallet<T> {
+			QueueCacheEntry { queue, total_size, consumed_count: 0, consumed_size: 0 }
+		})
+	}
+
+	/// Returns the message at the front of `para`'s queue, or `None` if the queue is empty.
 	///
-	/// Returns `(upward_message, became_empty)`, where
+	/// Does not mutate the queue.
+	fn peek_front<T: Config>(&mut self, para: ParaId) -> Option<&UpwardMessage> {
+		let entry = self.ensure_cached::<T>(para);
+		entry.queue.get(entry.consumed_count)
+	}
+
+	/// Attempts to remove one message from the front of `para`'s queue. If the queue is empty, then
+	/// does nothing.
 	///
-	/// - `upward_message` a dequeued message or `None` if the queue _was_ empty.
-	/// - `became_empty` is true if the queue _became_ empty.
-	fn dequeue<T: Config>(&mut self, para: ParaId) -> (Option<UpwardMessage>, bool) {
-		let cache_entry = self.0.entry(para).or_insert_with(|| {
-			let queue = <Pallet<T> as Store>::RelayDispatchQueues::get(&para);
-			let (count, total_size) = <Pallet<T> as Store>::RelayDispatchQueueSize::get(&para);
-			QueueCacheEntry { queue, count, total_size }
-		});
-		let upward_message = cache_entry.queue.pop_front();
-		if let Some(ref msg) = upward_message {
-			cache_entry.count -= 1;
-			cache_entry.total_size -= msg.len() as u32;
+	/// Returns `true` iff there are no more messages in the queue after the removal attempt.
+	fn consume_front<T: Config>(&mut self, para: ParaId) -> bool {
+		let cache_entry = self.ensure_cached::<T>(para);
+		let upward_message = cache_entry.queue.get(cache_entry.consumed_count);
+		if let Some(msg) = upward_message {
+			cache_entry.consumed_count += 1;
+			cache_entry.consumed_size += msg.len();
 		}
 
-		let became_empty = cache_entry.queue.is_empty();
-		(upward_message, became_empty)
+		cache_entry.consumed_count >= cache_entry.queue.len()
 	}
 
 	/// Flushes the updated queues into the storage.
 	fn flush<T: Config>(self) {
 		// NOTE we use an explicit method here instead of Drop impl because it has unwanted semantics
 		// within runtime. It is dangerous to use because of double-panics and flushing on a panic
 		// is not necessary as well.
-		for (para, QueueCacheEntry { queue, count, total_size }) in self.0 {
-			if queue.is_empty() {
+		for (para, entry) in self.0 {
+			if entry.consumed_count >= entry.queue.len() {
 				// remove the entries altogether.
-				<Pallet<T> as Store>::RelayDispatchQueues::remove(&para);
-				<Pallet<T> as Store>::RelayDispatchQueueSize::remove(&para);
-			} else {
-				<Pallet<T> as Store>::RelayDispatchQueues::insert(&para, queue);
-				<Pallet<T> as Store>::RelayDispatchQueueSize::insert(&para, (count, total_size));
+				RelayDispatchQueues::<T>::remove(&para);
+				RelayDispatchQueueSize::<T>::remove(&para);
+			} else if entry.consumed_count > 0 {
+				RelayDispatchQueues::<T>::insert(&para, &entry.queue[entry.consumed_count..]);
+				let count = (entry.queue.len() - entry.consumed_count) as u32;
+				let size = (entry.total_size as usize - entry.consumed_size) as u32;
```

Comment:
saturating math on specifically `total_size` may be safer, since there is nothing in the immediately surrounding code which ensures that this will not underflow